### PR TITLE
feat(linux): add function comments and remaining comments 🐘

### PR DESCRIPTION
**Pull request by SabineSIL**
fixes parts of PR https://github.com/keymanapp/keyman/pull/9384.

This is a PR for the remaining topics. Mostly formatting/editing things like 
- adding function comments
- format document

- replace codecvt
- set capslock =1
- replace Macro for calculating max
- prevent buffer overrun in u16tok 

- @keymanapp-test-bot skip

### CODE REVIEW

File: linux/mcompile/keymap/u16.h [hunk 1]
```diff
@@ -1,45 +1,201 @@
...
+ * @param  str the std::string to be converted
+ * @return a std::wstring
+ */
 std::wstring wstring_from_string(std::string const str);
-std::wstring wstring_from_u16string(std::u16string const str16);
+
+/**
+ * @brief  Obtain a std::string from a std::wstring
+ * @param  str the std::wstring to be converted
+ * @return a std::string
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.h:31]:
```suggestion
 * @brief  Obtain a std::u16string from a std::string
 * @param  str the std::string to be converted
 * @return a std::u16string
```

File: linux/mcompile/keymap/u16.h [hunk 2]
```diff
@@ -1,45 +1,201 @@
...
+ */
 std::u16string u16string_from_string(std::string const str);
-std::u16string u16string_from_wstring(std::wstring const wstr);
+
+/**
+ * @brief  Obtain a std::string from a std::u16string
+ * @param  str the std::string to be converted
+ * @return a std::u16string
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.h:38]:
```suggestion
 * @param  str the std::u16string to be converted
 * @return a std::string
```

File: linux/mcompile/keymap/u16.h [hunk 3]
```diff
@@ -1,45 +1,201 @@
...
+ * @return void
+ */
+void u16sprintf(KMX_WCHAR* dst, const size_t sz, const wchar_t* fmt, ...);
+
+/**
+ * @brief  Convert u16string to long integer
+ * @param  str u16string beginning with the representation of an integral number.
+ * @param  endptr Reference to the next character in str
+ * @param  base Numerical base (radix) that determines the valid characters and their interpretation
+ * @return a std::string
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.h:71]:
```suggestion
 * @return a long
```

File: linux/mcompile/keymap/u16.h [hunk 4]
```diff
@@ -1,45 +1,201 @@
...
+ * @param  src u16string to be appended
+ * @param  max Maximum number of characters to be appended.
+ * @return Pointer to the destination array
+ */
+const KMX_WCHAR* u16ncat(KMX_WCHAR* dst, const KMX_WCHAR* src, size_t max);
+
+/**
+ * @brief  Append a '/' or '\\' to an array of char16_t
+ * @param  Name Pointer to the source
+ * @return Pointer to the source with slash/backslash added
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.h:88]:
```suggestion
 * @brief  Find last '/' or '\\' in an array of char16_t
 * @param  Name Pointer to the source
 * @return Pointer to the last slash/backslash
```

File: linux/mcompile/keymap/u16.h [hunk 5]
```diff
@@ -1,45 +1,201 @@
...
+ */
 const KMX_WCHAR* u16rchr_slash(KMX_WCHAR const* Name);
 
+/**
+ * @brief  Append a '/' or '\\' to an array of char
+ * @param  Name Pointer to the source
+ * @return Pointer to the source with slash/backslash added
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.h:95]:
```suggestion
 * @brief  Find last '/' or '\\' in an array of char
 * @param  Name Pointer to the source
 * @return Pointer to the the last slash/backslash
```

File: linux/mcompile/keymap/u16.h [hunk 6]
```diff
@@ -1,45 +1,201 @@
...
+const KMX_WCHAR* u16chr(const KMX_WCHAR* p, KMX_WCHAR ch);
+
+/**
+ * @brief  Copy the u16string pointed by source into the array pointed by destination
+ * @param  dst Pointer to the destination
+ * @param  src Pointer to the source to be copied
+ * @return Pointer to the destination
+ */
+
+const KMX_WCHAR* u16cpy(KMX_WCHAR* dst, const KMX_WCHAR* src);
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.h:121]:
```suggestion
const KMX_WCHAR* u16cpy(KMX_WCHAR* dst, const KMX_WCHAR* src);

```

File: linux/mcompile/keymap/u16.cpp [hunk 7]
```diff
@@ -205,7 +186,8 @@ int   u16nicmp(const KMX_WCHAR *p, const KMX_WCHAR *q, size_t count) {
   return 0;
 }
 
-int   u16icmp(const KMX_WCHAR *p, const KMX_WCHAR *q) {
+/** @brief @brief Case sensitive comparison of two strings */
```
[the remark this answers is not in the document]
Comment (SabineSIL) [reply linux/mcompile/keymap/u16.cpp:189]: OK - done

File: linux/mcompile/keymap/keymap.h
[the remark this answers is not in the document]
Comment (SabineSIL) [reply linux/mcompile/keymap/keymap.h:1]: left it as is...

File: linux/mcompile/keymap/mc_kmxfile.cpp [hunk 8]
```diff
@@ -35,55 +41,82 @@ const int CODE__SIZE[] = {
...
+
+/**
+ * @brief Fixup the keyboard by expanding pointers. On disk the pointers are stored relative to the
+ *        beginning of the file, but we need real pointers. This method is used on 32-bit architectures.
+ * @param bufp pointer to buffer where data will be copied into
+ * @param base pointer to starting point
+ * @param dwFileSize size of the file
+ * @return pointer to the keyboard
+ */
+LPKMX_KEYBOARD KMX_FixupKeyboard(PKMX_BYTE bufp, PKMX_BYTE base, KMX_DWORD dwFileSize);
```
[the remark this answers is not in the document]
Comment (SabineSIL) [reply linux/mcompile/keymap/mc_kmxfile.cpp:61]: left it as is...

**Review by SabineSIL — COMMENTED** [review]
Thanks for reviewing. Here are the requested changes:

About putting detailed comments into the cpp-files:  
I asked earlier and Marc proposed the other way. Since it`s a lot of work to change, I leave it as he proposed for now. 


Marc wrote on slack: [Jul 3rd at 11:53 PM](https://keymanapp.slack.com/archives/C871HJXT7/p1720043619997529?thread_ts=1720022370.338909&cid=C871HJXT7): `Function comment goes in the .h file, with a one liner duplicate of the function purpose in the .cpp -- I think.`

**Comment by ermshiperete** [comment]
> About putting detailed comments into the cpp-files: I asked earlier and Marc proposed the other way. Since it`s a lot of work to change, I leave it as he proposed for now.

OK.

> Marc wrote on slack: [Jul 3rd at 11:53 PM](https://keymanapp.slack.com/archives/C871HJXT7/p1720043619997529?thread_ts=1720022370.338909&cid=C871HJXT7): `Function comment goes in the .h file, with a one liner duplicate of the function purpose in the .cpp -- I think.`

I don't fully agree. If it's for an API that is used externally then it makes sense to put the documentation in the .h file, because typically you have the binary library and the header file, so you look in the header file.

However, if it's only used internally, i.e. within the same project or repo, then I think you more often look at the definition in the .cpp file because not all functions have documentation comments and so then you're required to look at the definition to figure out the meaning of the parameters and what the function does. And if you use F12 in an IDE it'll take you to the definition if it's available. So IMO for internal functions it'd make more sense to have the full documentation in the .cpp file and the one-liner in the .h file.

File: linux/mcompile/keymap/deadkey.h [hunk 9]
```diff
@@ -7,66 +7,65 @@
...
 bool found_dk_inVector(KMX_WCHAR dk, std::vector<DeadKey*>& dkVec);
 
 /**
- * @brief find all deadkey combinations for a certain deadkey in a vector of all deadkey combinations
- * @param r_dk_ComposeTable vector containing all possible deadkey combinations
- * @param dk deadkey of interest
- * @param [in,out] dk_SingleTable vector containing all dk-character combinations for a specific deadkey dk
- * @return true if successful; false if not
+ * @brief  find all deadkey combinations for a certain deadkey in a vector of all deadkey combinations
+ * @param         r_dk_ComposeTable vector containing all possible deadkey combinations
```
Comment (ermshiperete) [inline linux/mcompile/keymap/deadkey.h:34]:
```suggestion
 * @param         dk_ComposeTable   vector containing all possible deadkey combinations
```

File: linux/mcompile/keymap/keymap.h [hunk 10]
```diff
@@ -103,52 +104,54 @@ bool ensureValidInputForKeyboardTranslation(int shiftstate, gint keycode);
...
 /**
- * @brief process each element of a 1D-Vector, split and write to a 3D-Vector
- * @param [in,out] all_US a 3D_Vector containing all keyvalues of the US keyboard
- * @param completeList a 1D-Vector containing all relevant entries copied from the symbol file us basic
- * @return 0 on success; 1 if entry can be split
+ * @brief  process each element of a 1D-Vector, split and write to a 3D-Vector
+ * @param[in,out] all_US        a 3D_Vector containing all keyvalues of the US keyboard
+ * @param         completeList  a 1D-Vector containing all relevant entries copied from the symbol file us basic
+ * @return 0 on success;
+ *         if entry can be split
```
Comment (ermshiperete) [inline linux/mcompile/keymap/keymap.h:154]: ?

File: linux/mcompile/keymap/keymap.h [hunk 11]
```diff
@@ -161,21 +164,22 @@ int split_US_To_3D_Vector(vec_dword_3D& all_US, vec_string_1D completeList);
...
  */
 int append_underlying_ToVector(vec_dword_3D& all_vector, GdkKeymap* keymap);
 
 /**
  * @brief  create a pointer to pointer of the current keymap for later use
- * @param  keymap  pointer to pointer to currently used (underlying) keyborad layout
- * @param  argc count of arguments
- * @param  argv array of arguments
- * @return 0 on success; 1 if the display is not found; 2 if the keymap is not found
+ * @param  keymap pointer to pointer to currently used (underlying) keyborad layout
```
Comment (ermshiperete) [inline linux/mcompile/keymap/keymap.h:178]:
```suggestion
 * @param  keymap pointer to pointer to currently used (underlying) keyboard layout
```

File: linux/mcompile/keymap/keymap.h [hunk 12]
```diff
@@ -644,26 +650,26 @@ KMX_DWORD KMX_get_KeyValUnderlying_From_KeyValUS(vec_dword_3D& all_vector, KMX_D
...
+ * @param  ss         a windows-type shiftstate
+ * @param  caps       state of the caps key
  * @return the keycode of the underlying keyboard if found;
  * 				 else the keycode of the US keyboard
  */
 KMX_DWORD KMX_get_KeyCodeUnderlying_From_KeyCodeUS(GdkKeymap* keymap, vec_dword_3D& all_vector, KMX_DWORD kc_us, ShiftState ss, int caps);
 
 /**
- * @brief  return the keycode of the currently used (underlying) keyboard for a given virtual key of the US keyboard
+ *  return the keycode of the currently used (underlying) keyboard for a given virtual key of the US keyboard
```
Comment (ermshiperete) [inline linux/mcompile/keymap/keymap.h:664]: That's missing the `@brief`

File: linux/mcompile/keymap/keymap.h [hunk 13]
```diff
@@ -644,26 +650,26 @@ KMX_DWORD KMX_get_KeyValUnderlying_From_KeyValUS(vec_dword_3D& all_vector, KMX_D
...
  *         "Where on an underlying keyboard do we find a character of a US keyboard?"
  * @param  virtualKeyUS a virtual key of the US keyboard
  * @return the keycode of the currently used (underlying) keyboard
  */
 KMX_DWORD KMX_get_KeyCodeUnderlying_From_VKUS(KMX_DWORD virtualKeyUS);
 
 /**
- * @brief  return a virtual key of the US keyboard for a given keycode of the currently used (underlying) keyboard
+ *  return a virtual key of the US keyboard for a given keycode of the currently used (underlying) keyboard
```
Comment (ermshiperete) [inline linux/mcompile/keymap/keymap.h:672]: That's missing the `@brief`

File: linux/mcompile/keymap/u16.cpp [hunk 14]
```diff
@@ -23,35 +24,31 @@ std::wstring wstring_from_string(std::string const str) {
...
+/** @brief Obtain a std::u16string from a std::wstring */
+std::u16string u16string_from_wstring(std::wstring const wstr) {
+  return convert<wchar_t, char16_t>((const std::wstring)wstr);;
 }
 
-  // UTF16 (=const wchar_t*) -> -> std::string  -> std::u16string -> UTF16 ( = char16_t*)
-/** @brief @brief  Convert pointer to wchar_t to u16string and copy sz elements into dst */
+// UTF16 (=const wchar_t*) -> -> std::string  -> std::u16string -> UTF16 ( = char16_t*)
+/** @brief Convert pointer to wchar_t to u16string and copy sz elements into dst */
+
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.cpp:51]:
```suggestion
```

File: linux/mcompile/keymap/u16.h [hunk 15]
```diff
@@ -145,60 +136,60 @@ size_t  u16len(const KMX_WCHAR* p);
...
- * @brief Split u16string into tokens
- * @param  p Pointer to u16string to truncate.
- * @param  ch the delimiter character
- * @param  ctx the string until and without the delimiter
- * @return Pointer to the destination or NULL
+ * @brief  Split u16string into tokens
+ * @param  p   Pointer to u16string to parse.
+ * @param  ch  the delimiter character
+ * @param  ctx the remaining string after the first delimiter
+ * @return PPointer to the first token in p
```
Comment (ermshiperete) [inline linux/mcompile/keymap/u16.h:178]:
```suggestion
 * @return Pointer to the first token in p
```

**Review by ermshiperete — CHANGES_REQUESTED** [review]
Almost there

File: linux/mcompile/keymap/deadkey.h (hunk 9 above)
Comment (SabineSIL) [reply linux/mcompile/keymap/deadkey.h:34]: OK-done (changed to r_dk_ComposeTable)

**Review by SabineSIL — COMMENTED** [review]
done all changes except for position of documentation

**Comment by SabineSIL** [comment]
> > About putting detailed comments into the cpp-files: I asked earlier and Marc proposed the other way. Since it`s a lot of work to change, I leave it as he proposed for now.
> 
> OK.
> 
> > Marc wrote on slack: [Jul 3rd at 11:53 PM](https://keymanapp.slack.com/archives/C871HJXT7/p1720043619997529?thread_ts=1720022370.338909&cid=C871HJXT7): `Function comment goes in the .h file, with a one liner duplicate of the function purpose in the .cpp -- I think.`
> 
> I don't fully agree. If it's for an API that is used externally then it makes sense to put the documentation in the .h file, because typically you have the binary library and the header file, so you look in the header file.
> 
> However, if it's only used internally, i.e. within the same project or repo, then I think you more often look at the definition in the .cpp file because not all functions have documentation comments and so then you're required to look at the definition to figure out the meaning of the parameters and what the function does. And if you use F12 in an IDE it'll take you to the definition if it's available. So IMO for internal functions it'd make more sense to have the full documentation in the .cpp file and the one-liner in the .h file.

@mcdurdin : What do you think? How do we want to do that?

File: linux/mcompile/keymap/mcompile.cpp [hunk 16]
```diff
@@ -70,6 +47,16 @@ std::vector<KMX_DeadkeyMapping> KMX_FDeadkeys;  // I4353
   run(argc, argv);
 }
 
+
+/**
+ * @brief  start of mcompile; load, convert and save keyboard
+ * @param  argc    number of commandline arguments
+ * @param  argv     pointer to commandline arguments: executable, inputfile, outputfile
+ * @param  argv_gdk pointer to (commandline arguments)
```
Comment (ermshiperete) [inline linux/mcompile/keymap/mcompile.cpp:55]: There's no `argv_gdk` parameter...

```suggestion
 * @param  argv     pointer to commandline arguments: executable, inputfile, outputfile
```

**Review by ermshiperete — APPROVED** [review]
A few minor things left...

**Comment by SabineSIL** [comment]
changed minor things

Comments:
- detailed comment in .cpp   -> before function definition
- one-liner comment  in .cpp -> before forward declaration
- one-liner comment  in .h     -> before declaration

**Pull request merged**